src/Microsoft.ServiceHub.Framework/AsyncNamedPipeClientStream.cs

```diff
@@ -0,0 +1,174 @@
+			PipeOptions options,
+			TokenImpersonationLevel impersonationLevel = TokenImpersonationLevel.None)
+			: base(direction, 4096)
+		{
+			Requires.NotNullOrEmpty(serverName, nameof(serverName));
+			Requires.NotNullOrEmpty(pipeName, nameof(pipeName));
+
+			this.pipePath = $@"\\{serverName}\pipe\{pipeName}";
+			this.direction = direction;
+			this.pipeOptions = options;
+			this.impersonationLevel = impersonationLevel;
+			this.access = this.GetAccess();
+		}
+
+		/// <summary>
+		/// Connects pipe client to server.
+		/// </summary>
+		/// <param name="cancellationToken">Cancellation token.</param>
+		/// <param name="maxRetries">Maximum number of retries.</param>
+		/// <param name="retryDelayMs">Milliseconds delay between retries.</param>
+		/// <returns>A task representing the establishment of the client connection.</returns>
+		/// <exception cref="TimeoutException">Thrown if no connection can be established.</exception>
+		public async Task ConnectAsync(
+			CancellationToken cancellationToken,
+			int maxRetries,
+			int retryDelayMs)
+		{
+			var errorCodeMap = new Dictionary<int, int>();
+			int retries = 0;
+			while (!cancellationToken.IsCancellationRequested)
+			{
+				if (retries > maxRetries || this.TryConnect(ref errorCodeMap))
+				{
+					break;
+				}
+
+				retries++;
+				await Task.Delay(retryDelayMs, cancellationToken).ConfigureAwait(false);
+			}
+
+			if (retries > maxRetries || !this.IsConnected)
+			{
+				throw new TimeoutException($"Failed with errors: {string.Join(", ", errorCodeMap.Select(x => $"(code: {x.Key}, count: {x.Value})"))}");
+			}
+		}
+
+		[DllImport("kernel32.dll", EntryPoint = "CreateFileW", CharSet = CharSet.Unicode, SetLastError = true)]
+		private static extern SafePipeHandle CreateNamedPipeClient(
+			string lpFileName,
+			uint dwDesiredAccess,
+			uint dwShareMode,
+			IntPtr secAttrs,
+			uint dwCreationDisposition,
+			uint dwFlagsAndAttributes,
+			IntPtr hTemplateFile);
+
+		private bool TryConnect(ref Dictionary<int, int> errorCodeMap)
+		{
+			var pipeFlags = this.GetPipeFlags();
+			SafePipeHandle handle = CreateNamedPipeClient(this.pipePath, (uint)this.access, (uint)FileShare.None, IntPtr.Zero, (uint)FileMode.Open, (uint)pipeFlags, IntPtr.Zero);
+
+			if (handle.IsInvalid)
+			{
+				int errorCode = Marshal.GetLastWin32Error();
+				handle.Dispose();
+				var newErrorCount = errorCodeMap.ContainsKey(errorCode) ? errorCodeMap[errorCode] + 1 : 1;
+				errorCodeMap[errorCode] = newErrorCount;
+				return false;
+			}
+
+			// Success!
+			this.InitializeHandle(handle, false, true);
+			this.IsConnected = true;
+			this.ValidateRemotePipeUser();
+			return true;
+		}
+
+		private void ValidateRemotePipeUser()
+		{
+#if !NETSTANDARD
+#if NETFRAMEWORK
+			var isCurrentUser = (this.pipeOptions & ~PolyfillExtensions.PipeOptionsCurrentUserOnly) != 0;
+#else
+			var isCurrentUser = (this.pipeOptions & ~PipeOptions.CurrentUserOnly) != 0;
+#endif
+
+			if (isCurrentUser)
```

Comment:
🤔 Why are we returning at this point? It appears that the only check so far is regarding whether 'current user only' was a _desired_ state, but we haven't yet verified that it _is_ the actual state.